File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> hash(256, -1);
        int maxi=0;
        int l=0;
        int r=0;
        int n=s.size();
        while(r<n){
            if(hash[s[r]]!=-1){
                if(hash[s[r]]>=l){
                    l=hash[s[r]]+1;
                }
            }
            maxi=max(maxi, r-l+1);
            hash[s[r]]=r;
            r++;
        }
        return maxi;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> hash=new HashMap<>();
        int n=s.length();
        int i=0;
        int e=0;
        int maxi=0;
        while(e<n){
            if(hash.containsKey(s.charAt(e))){
                if(hash.get(s.charAt(e))>=i)
                    i=hash.get(s.charAt(e))+1;
            }
            maxi=Math.max(maxi, e-i+1);
            hash.put(s.charAt(e),e);
            e++;
        }
        return maxi;
    }
}